### DISCUSSION TITLE
Is it possible to generate type definitions with this package?

### DISCUSSION THREAD

**Discussion by mechadragon01** [Swatinem/rollup-plugin-dts#207] [Q&A]
I am looking at the package description for [rollup-plugin-swc3](https://www.npmjs.com/package/rollup-plugin-swc3) 

"Use rollup-plugin-dts which generates and bundle .d.ts, also does type checking. It is used by this plugin as well."

I can only get this plugin to work when `tsc` generates the type definition files. How do I get rollup-plugin-dts to generate type definition files?

**Comment by Swatinem** [comment]
You can just give it the `.ts` file as input, and it will internally generate the `.d.ts` for you. Though usage of this mode is discouraged.

**Reply by linonetwo** [reply]
Why, is it slow? But generating .d.ts file using typescript then feed into this plugin will be slower, is it?

**Reply by Swatinem** [reply]
I believe its about the same speed, as the typescript compiler will need to do the type checking, inference and definition generation either way.
It is also a huge maintenance burden.

**Reply by mesqueeb** [reply]
@Swatinem 
> You can just give it the .ts file as input, and it will internally generate the .d.ts for you. Though usage of this mode is discouraged.

Why is this sort of usage discouraged?



### using this plugin to generate `.d.ts` files from `.ts` files
It's the only way I found to easily do this via a rollup plugin where we have the ability to explicitly state the export location & filename of the declaration file.

Other rollup plugins also encourage usage in this way, like: https://github.com/egoist/rollup-plugin-esbuild#declaration-file

### other options
I find that besides using `tsc` this is one of the only rollup plugins that properly generates the .d.ts files... (another option is to use `rollup-plugin-typescript2` for generating the declaration files, but with this plugin it's impossible to explicitly state a single end-point for your declaration files, nor deciding what those files should use as filename... 🤦🏻‍♂️)

### the conundrum
In short, this plugin is purrfect!
But if you say it _shouldn't_ be used in this way, I'm at a loss how to best generate the declaration files with rollup. 🤔  Any advice or insights?

**Reply by Swatinem** [reply]
The reason I discourage it is that it is a huge maintenance liability.

Making sure that declarations are rolled-up is the main purpose of this plugin.
Generating declarations from full TS code is not its main purpose, but takes up a consirable amount of code to get going, and its the primary thing people have problems with, as there is a ton of things to have to get right, and to consider. And a ton of options and permutations to deal with.

**Reply by linonetwo** [reply]
> I find that besides using tsc this is one of the only rollup plugins that properly generates the .d.ts files

I agree with this.

So is there a guide to generate a d.ts file of codebase? Run tsc first, and ask it to output things to /dist/types folder, then use this plugin to aggregate those generated .d.ts file?

**Reply by StreetStrider** [reply]
After looking through some rollup — TS related packages and reading through this discussion I became more and more confused.

Right now I'm experimenting on prepping libraries with Rollup. I want to generate `TS → JS ESM + d.ts` on 1-to-1 basis for library use-case.

I can generate typings with `tsc … --declaration --emitDeclarationOnly …` (it worked), but I looked for a way that is more integrated with Rollup. So I've tried to generate `.d.ts` from `.ts` with this package, it worked to some extent, but some notes in the readme and this thread says that this is not the use-case for this library.

Is there any package dedicated to `.ts → .d.ts` generation in Rollup world?

**Reply by linonetwo** [reply]
I'm now using tsc, this is my setup for reference: https://github.com/tiddly-gittly/wikiast/blob/1c19d50702bc373a94bcb4778540d45977c9339a/packages/wikiast-util-to-wikitext/package.json#L9-L10

**Reply by StreetStrider** [reply]
@linonetwo thanks, I'm thinking of switching to the same logic and run it in parallel with Rollup. For the moment I have no clear vision on how to implement this with rollup + some ts plugin on my own, nor any help from typescript-related plugins' maintainers.
